packages/@aws-cdk/aws-kinesisfirehose/lib/destination.ts

```diff
@@ -0,0 +1,99 @@
+import * as logs from '@aws-cdk/aws-logs';
+import { Construct } from 'constructs';
+import { IDeliveryStream } from './delivery-stream';
+import { CfnDeliveryStream } from './kinesisfirehose.generated';
+
+/**
+ * A Kinesis Data Firehose delivery stream destination configuration.
+ */
+export interface DestinationConfig {
+  /**
+   * Schema-less properties that will be injected directly into `CfnDeliveryStream`.
+   */
+  readonly properties: object;
+}
+
+/**
+ * Options when binding a destination to a delivery stream.
+ */
+export interface DestinationBindOptions {
+  /**
+   * The delivery stream.
+   */
+  readonly deliveryStream: IDeliveryStream;
+}
+
+/**
+ * A Kinesis Data Firehose delivery stream destination.
+ */
+export interface IDestination {
+  /**
+   * Binds this destination to the Kinesis Data Firehose delivery stream.
+   *
+   * Implementers should use this method to bind resources to the stack and initialize values using the provided stream.
+   */
+  bind(scope: Construct, options: DestinationBindOptions): DestinationConfig;
+}
+
+/**
+ * Generic properties for defining a delivery stream destination.
+ */
+export interface DestinationProps {
+  /**
+   * If true, log errors when data transformation or data delivery fails.
+   *
+   * If `logGroup` is provided, this will be implicitly set to `true`.
+   *
+   * @default true - errors are logged.
+   */
+  readonly logging?: boolean;
+
+  /**
+   * The CloudWatch log group where log streams will be created to hold error logs.
+   *
+   * @default - if `logging` is set to `true`, a log group will be created for you.
+   */
+  readonly logGroup?: logs.ILogGroup;
+}
+
+/**
+ * Abstract base class that destination types can extend to benefit from methods that create generic configuration.
+ */
+export abstract class DestinationBase implements IDestination {
+  private logGroups: { [logGroupId: string]: logs.ILogGroup } = {};
+
+  constructor(protected readonly props: DestinationProps = {}) { }
```

Comment:
I don't think it's a great idea to make the `props` supertype a `protected` attribute. It's not a lot of effort to have every subclass make the more specific type `private`, and it'll be more accurate.

Reply:
So would that mean that everything in `props` that gets used by the helper functions i.e. createLoggingOptions, would need to be passed in instead? 

That adds some code, but not a lot. I'm okay with that change. 

Curious, what is the reason why it's not good to have `props` as a `protected` attribute?